homeassistant/components/device_tracker/owntracks_http.py

```diff
@@ -4,52 +4,83 @@
 
-DEPENDENCIES = ['http']
+EVENT_RECEIVED = 'owntracks_http_webhook_received'
+EVENT_RESPONSE = 'owntracks_http_webhook_response_'
 
+DOMAIN = 'device_tracker.owntracks_http'
 
-async def async_setup_scanner(hass, config, async_see, discovery_info=None):
-    """Set up an OwnTracks tracker."""
-    context = context_from_config(async_see, config)
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_WEBHOOK_ID): cv.string
+})
 
-    hass.http.register_view(OwnTracksView(context))
 
+async def async_setup(hass, config):
+    """Set up the OwnTracks HTTP service component."""
     return True
 
 
-class OwnTracksView(HomeAssistantView):
-    """View to handle OwnTracks HTTP requests."""
-
-    url = '/api/owntracks/{user}/{device}'
-    name = 'api:owntracks'
-
-    def __init__(self, context):
-        """Initialize OwnTracks URL endpoints."""
-        self.context = context
-
-    async def post(self, request, user, device):
-        """Handle an OwnTracks message."""
-        hass = request.app['hass']
+async def async_setup_scanner(hass, config, async_see, discovery_info=None):
+    """Set up OwnTracks HTTP component."""
+    context = context_from_config(async_see, config)
 
-        subscription = self.context.mqtt_topic
-        topic = re.sub('/#$', '', subscription)
+    subscription = context.mqtt_topic
+    topic = re.sub('/#$', '', subscription)
+
+    async def handle_webhook(hass, webhook_id, request):
+        """Handle webhook callback."""
+        headers = request.headers
+        data = dict()
+
+        if 'X-Limit-U' in headers:
+            data['user'] = headers['X-Limit-U']
+        elif 'u' in request.query:
+            data['user'] = request.query['u']
+        else:
+            return Response(
+                body=json.dumps({'error': 'You need to supply username.'}),
+                content_type="application/json"
+            )
+
+        if 'X-Limit-D' in headers:
+            data['device'] = headers['X-Limit-D']
+        elif 'd' in request.query:
+            data['device'] = request.query['d']
+        else:
+            return Response(
+                body=json.dumps({'error': 'You need to supply device name.'}),
+                content_type="application/json"
+            )
 
         message = await request.json()
-        message['topic'] = '{}/{}/{}'.format(topic, user, device)
 
-        try:
-            await async_handle_message(hass, self.context, message)
-            return self.json([])
+        message['topic'] = '{}/{}/{}'.format(topic, data['user'],
+                                             data['device'])
 
+        try:
+            await async_handle_message(hass, context, message)
+            return Response(body=json.dumps([]), status=200,
+                            content_type="application/json")
         except ValueError:
-            raise HTTPInternalServerError
+            return Response(body=json.dumps({'error': 'Error parsing JSON'}),
```

Comment:
We should log an error but just return `None`. Webhooks are unauthenticated and we don't want to give away that they hit a valid webhook.

Reply:
As a matter of fact, if noone specifies u=USERNAME&d=DEVICE as part of the webhook URL this line would never be hit, instead the "Provide username" error will be returned. I've adjusted it nevertheless.

The provide username error is cryptic enough and doesn't reveal that the username should be provided as part of the webhook query string.

Reply:
good point.